### PR TITLE
User account registration

### LINKED ISSUE #1: User account registration (server)
> As a student,
So I can join the network,
I want to create an account

**Requirements**
- POST endpoint `/user`
- Password must be hashed before stored

**Request body**
```json
{
  "first_name": string,
  "last_name": string,
  "email": string,
  "password": string,
  "biography": string | null,
  "github_url": string | null
}
```

**201 Response body**
```json
{
  "status": "success",
  "data": {
    "user": {
      "id": int,
      "cohort_id": int | null,
      "first_name": string,
      "last_name": string,
      "email": string,
      "role": string<"STUDENT", "TEACHER">,
      "biography": string | null,
      "github_url": string | null
    }
  }
}
```

**4xx Client error response body**
```json
{
  "status" : "fail",
  "data" : {
    "first_name": "First name is required",
    "password": "Password must be at least 8 characters"
  }
}
```

**5xx Server error response body**
```json
{
  "status": "error",
  "message": "Unable to communicate with database"
}
```

### PR DESCRIPTION
**Pull request by vherus**
Closes #1

### CODE REVIEW

File: src/domain/User.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    this.id = id
+    this.firstName = firstName
+    this.lastName = lastName
+    this.email = email
+    this.bio = bio
+    this.githubUrl = githubUrl
+    this.passwordHash = passwordHash
+  }
+
+  static async fromJson(json) {
```
[the remark this answers is not in the document]
Comment (Auenc) [reply src/domain/User.js:49]: Just thinking that it's not something that's been shown to the students, so they might be confused as to what is happening 🤔

**Review by Auenc — APPROVED** [review]

File: src/domain/User.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (vherus) [reply src/domain/User.js:49]: I think this might be a good move

File: src/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { User, createNewUser } from '../domain/user.js'
+
+export const create = async (req, res) => {
+  const userToCreate = await User.fromJson(req.body)
+
+  try {
+    const createdUser = await createNewUser(userToCreate)
```
Comment (dearshrewdwit) [inline src/controllers/user.js:7]: What about leverage the class and defining a `User#save()` method?

```js
 userToCreate.save()
```

File: src/domain/User.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        first_name: this.firstName,
+        last_name: this.lastName,
+        email: this.email,
+        biography: this.bio,
+        github_url: this.githubUrl
+      }
+    }
+  }
+
+  static fromDb(user) {
```
Comment (dearshrewdwit) [inline src/domain/User.js]: let's move static methods to the top of class definitions

File: src/controllers/user.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    res.status(201).json({
+      status: 'success',
+      data: createdUser
+    })
+  } catch (error) {
+    res.status(500).json({
+      status: 'error',
+      message: 'Unable to create new user'
```
[the remark this answers is not in the document]
Comment (dearshrewdwit) [reply src/controllers/user.js:16]: Probably - ticket this up and add it for students later

**Pull request merged**